# __SPARK ASSIGNMENT 1__
> Name : Ilias Dimos <br />
> AM : f2822102 <br />
> Course : Big Data Systems <br />
> Email : f2822102@aueb.gr <br />

## __TASK 1__
#### Your first task is to explore the dataset. You need to use SparkSQL with Dataframes in a Jupyter notebook that delivers the following : 
#### __1) It uses the json() function to load the dataset.__

In [20]:
from pyspark.sql import SparkSession

#Builing the Spark session with name "Assignment_1_TASK_1
spark = SparkSession.builder.appName("Assignment_1_TASK1_1").getOrCreate()

#Importing the "books_5000.json" file with the read.json() command.
dataset = spark.read.json("books_5000.json")

#Showing the first 5 lines of our dataset
dataset.show(5)

+----------+--------------------+--------------+--------+------------+--------------------+-------------------+---------+--------------------+--------+----------+-------------+-----------+-------------+--------------------+---------+--------------------+---------------+-----------------+----------------+--------------------+-------------+--------------------+--------------------+------------------+--------------------+--------------------+--------------------+--------+
|      asin|             authors|average_rating| book_id|country_code|         description|edition_information|   format|           image_url|is_ebook|      isbn|       isbn13|kindle_asin|language_code|                link|num_pages|     popular_shelves|publication_day|publication_month|publication_year|           publisher|ratings_count|              series|       similar_books|text_reviews_count|               title|title_without_series|                 url| work_id|
+----------+--------------------+--------------+----

#### __2) It counts and displays the number of books in the database.__
* In total we have __4999__ books in the database.

In [21]:
dataset.count()

4999

In [22]:
dataset.distinct().count() # no duplicates

4999

#### __3) It counts and displays the number of e-books in the database (based on the “is_ebook” field).__
* In this question i choosed the `filter` function in order to extract the column that `contains` , if a book is `e_book` or not , and `count` how many of them really are. 
* If a book is e_book is labeled as `true` in the dataset.
* The Ebooks in the database are __749.__

In [23]:
from pyspark.sql.functions import col
dataset.filter(col("is_ebook").contains("true")).count()

749

#### __4) It uses the summary() command to display basic statistics about the “average_rating” field.__
* By seeing the `summary` table below we can see that we have 4999 `ratings` in our database.The `average rating` of the books is __3.91120.__ 
* The `Minimum rating` is 1 and the `Maximum rating` is __5.__


In [24]:
dataset.select("average_rating").summary().show()

+-------+-------------------+
|summary|     average_rating|
+-------+-------------------+
|  count|               4999|
|   mean| 3.9112042408481678|
| stddev|0.43444489528688784|
|    min|               1.00|
|    25%|               3.66|
|    50%|               3.98|
|    75%|               4.23|
|    max|               5.00|
+-------+-------------------+



#### __4) It uses the groupby() and count() commands to display all distinct values in the “format” field and their number of appearances.__
* We see that the `format` with the most appearances in the dataset is the "Paperback" with __2629.__

In [25]:
dataset_format_ordered = dataset.groupby("format").count().orderBy('count', ascending=False)
dataset_format_ordered.show(truncate=False)

+--------------------------+-----+
|format                    |count|
+--------------------------+-----+
|Paperback                 |2629 |
|                          |1245 |
|Hardcover                 |826  |
|ebook                     |84   |
|Mass Market Paperback     |64   |
|Kindle Edition            |41   |
|Comic Book                |15   |
|Comic                     |15   |
|Board book                |11   |
|Unknown Binding           |7    |
|Bolsillo con sobrecubierta|2    |
|Webtoon                   |2    |
|paperback                 |2    |
|Paperback Manga           |2    |
|Broche                    |2    |
|Library Binding           |2    |
|Illustrated               |2    |
|Comics                    |2    |
|Graphic Novel             |2    |
|Album                     |2    |
+--------------------------+-----+
only showing top 20 rows

